### PR TITLE
fix shpOptions.getGeometry()

### PR DESCRIPTION
**Pull request by simei94**
If a shpFile had multiple features, the getGeometry method of class ShpOptions did not include all features into the geometry, but only the first one because when iterating over the shpFile's features the features were collected but the geometry was NOT OVERWRITTEN, which resulted in the geometry just containing the shpFile's first feature.

### CODE REVIEW

File: contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java [hunk 1]
```diff
@@ -61,4 +63,35 @@ public void all() {
 				.hasSize(Set.copyOf(ft).size());
 
 	}
+
+	@Test
+	public void testGetGeometry() {
```
Comment (Janekdererste) [inline contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java:68]: I found it quite difficult to understand the intend of this test. Naming the variables geometry1,2,3 makes it very hard to understand what the purpose of those variables are. I think a better naming for `geometry1` would be `joinedResult` and for `geometry2` `expectedResult` for example. Of course you can choose any other name, which describes what these things are for. 

I think the comparison of `geometry1` and `geometry3` doesn't make sense. You want to test whether this method correctly joins all the geometries. This would be much clearer without `geometry3`. I can see why you have it in there but the test would also fail if only the geometry of the first feature was returned, as it was the case before your fix. 

Of course one test is better than no test! Nevertheless, I can see four execution paths in the `getGeometry` method. If you want to do it right you could add tests for:

- No features - Exception is expected
- One feature - This could be done the way you handled `geometry3`
- Multiple features - Some feature without geometry 
- Multiple features - Everything is as expected -> The happy path (your test)

File: contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java [hunk 2]
```diff
@@ -61,4 +63,35 @@ public void all() {
...
+
+		ShpOptions shp = new ShpOptions(input, null, null);
+		Geometry geometry1 = shp.getGeometry() ;
+		Geometry geometry2 = null ;
+		Geometry geometry3 = null ;
+
+		List<SimpleFeature> features = shp.readFeatures();
+
+		geometry3 = (Geometry) features.get(0).getDefaultGeometry();
+
```
Comment (Janekdererste) [inline contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java:85]: The way you have structured this for loop was also hard to parse for me. I would rather initialize a geometry before the loop and then have no if/else statement in the loop
```
		var geometry = new GeometryFactory().createEmpty(2);
		
		for (var feature : features) {
			var geometryToJoin = (Geometry) feature.getDefaultGeometry();
			geometry = geometry.union(geometryToJoin);
		}
```

**Review by Janekdererste — APPROVED** [review]
This works as is. If you're motivated you can do some of the changes I suggested.

**Comment by Janekdererste** [comment]
Fixes #2003

File: contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java [hunk 3]
```diff
@@ -61,4 +63,35 @@ public void all() {
...
+		for(SimpleFeature feature : features) {
+			if(geometry2 == null) {
+				geometry2 = (Geometry) feature.getDefaultGeometry();
+			} else {
+				geometry2 = geometry2.union((Geometry) feature.getDefaultGeometry());
+			}
+		}
+
+		Assert.assertFalse(geometry1.equals(geometry3));
```
[the remark this answers is not in the document]
Comment (simei94) [reply contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java:94]: I figured that it's better to use the geometry's own .equals method as they are explicitely comparing geometries and not just objects. Also, if I use Assert.assertEquals this test will fail. I think this has to do with the way the method .getGeometry and .readFeatures read in the features. Both now produce the same .shp, but if you look closely on their structure, the readFeatures method produces some brackets, which .getGeometry does not produce. Because of this the result of the functions is not EXACTLY the same, which does not matter for the resulting shape, so I think it's fine.

File: contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java (hunk 1 above)
Comment (simei94) [reply contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java:68]: I did some renaming and deleted gemeotry3. For now I won't add further tests, as I have lots of other stuff to do.

File: contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java (hunk 2 above)
Comment (simei94) [reply contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java:85]: re-structured the for loop, thanks for the advice

File: contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Janekdererste) [reply contribs/application/src/test/java/org/matsim/application/options/ShpOptionsTest.java]: Ah, I see. That makes sense.

**Comment by Janekdererste** [comment]
This is much easier to read for me now. Very nice!
